packages/x-components/src/composables/create-use-device.composable.ts

```diff
@@ -0,0 +1,206 @@
+  Ref<boolean>
+> &
+  Record<`is${Capitalize<Device>}OrGreater`, Ref<boolean>> &
+  Record<`is${Capitalize<Device>}OrLess`, Ref<boolean>>;
+
+/**
+ * Factory function that creates a composable for device detection using the devices parameter
+ * to configure breakpoints.
+ *
+ * @param devices - An object containing the breakpoints, where the key is the name of the device
+ * and the value is the screen width.
+ * @returns A composable which provides multiple reactive flags and values for detecting the
+ * current device. The flags names depends on the names passed in the `devices` parameter.
+ * @remarks The `orientation` only works for orientation- sensor devices (mobile, tablet, etc). If
+ * in a desktop, the height of the window is larger than the width, the orientation will be
+ * `landscape`.
+ *
+ * @example
+ * ´´´typescript
+ * const useDevice = createUseDevice(\{ mobile: 0, tablet: 744, desktop: 1024 \});
+ * const \{
+ *   isMobile,
+ *   isMobileOrLess,
+ *   isMobileOrGreater,
+ *   isTablet,
+ *   isTabletOrLess,
+ *   isTabletOrGreater,
+ *   isDesktop,
+ *   isDesktopOrLess,
+ *   isDesktopOrGreater,
+ *   deviceName,
+ *   orientation,
+ *   istTouchable
+ * \} = useDevice();
+ *
+ * @public
+ */
+export function createUseDevice<Device extends string>(
+  devices: Record<Device, number>
+): () => UseDeviceReturn<Device> {
+  let devicesFlags: UseDeviceFlags<Device>;
+  let orientation: UseDeviceReturn['orientation'];
+  let isTouchable: UseDeviceReturn['isTouchable'];
+  let deviceName: UseDeviceReturn['deviceName'];
+  // The `effectScope` group all the changes in one to avoid multiple re-renderings.
+  const scope = effectScope();
+  scope.run(() => {
+    devicesFlags = getDeviceFlags(devices);
+    orientation = getOrientation();
+    isTouchable = getIsTouchable();
+    deviceName = getDeviceName(devices, devicesFlags);
+  });
+  return () => ({
+    ...devicesFlags,
+    orientation,
+    isTouchable,
+    deviceName
+  });
+}
+
+/**
+ * A function that returns reactive flags to detect the current device based on provided
+ * breakpoints.
+ *
+ * @param devices - An object containing the breakpoints, where the key is the name of the device
+ * and the value is the screen width.
+ * @returns A object containing the multiple reactive flags.
+ *
+ * @internal
+ */
+function getDeviceFlags<Device extends string>(
+  devices: Record<Device, number>
+): UseDeviceFlags<Device> {
+  const { width: windowSize } = useWindowSize();
+  const sortedByWidthDevices = Object.entries<number>(devices).sort(
+    ([, aWidth], [, bWidth]) => bWidth - aWidth
+  ) as Array<[Device, number]>;
+  return reduce(
+    devices,
+    (accumulator, device, deviceWidth) => {
+      const isDevice = computed(
+        () =>
+          deviceWidth <= windowSize.value &&
+          !sortedByWidthDevices.some(
+            ([, otherDeviceWidth]) =>
+              otherDeviceWidth <= windowSize.value && otherDeviceWidth > deviceWidth
+          )
```

Comment:
Can we extract this to its own function? I feel like `getDeviceFlags` does a lot already

Reply:
done